compiler/src/iree/compiler/Codegen/Common/FlattenMemRefSubspanPass.cpp

```diff
@@ -665,6 +665,135 @@ struct FoldSubspanOffsetIntoLoadStore final : public OpRewritePattern<OpType> {
+struct FoldSubspanOffsetIntoMMALoad final
+    : public OpRewritePattern<gpu::SubgroupMmaLoadMatrixOp> {
+  using OpRewritePattern<gpu::SubgroupMmaLoadMatrixOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(gpu::SubgroupMmaLoadMatrixOp op,
+                                PatternRewriter &rewriter) const override {
+    auto memrefType = op.getSrcMemref().getType().template cast<MemRefType>();
+    if (!isRankOneMemRef(memrefType)) {
+      return rewriter.notifyMatchFailure(op, "expected 1-D memref");
+    }
+
+    auto subspanOp =
+        op.getSrcMemref()
+            .template getDefiningOp<IREE::HAL::InterfaceBindingSubspanOp>();
+    if (!subspanOp) return failure();
+
+    // If the subspan op has a zero byte offset then we are done.
+    if (!subspanOp.getByteOffset() ||
+        matchPattern(subspanOp.getByteOffset(), m_Zero())) {
+      return failure();
+    }
+
+    // Calculate the offset we need to add to the mma load op, in terms of how
+    // many elements.
+    Optional<int64_t> numBytes = getNumBytes(memrefType.getElementType());
+    if (!numBytes) {
+      return rewriter.notifyMatchFailure(op,
+                                         "cannot deduce element byte count");
+    }
+    // Create a new subspan op with zero byte offset at the original location.
+    auto ip = rewriter.saveInsertionPoint();
+    rewriter.setInsertionPointAfter(subspanOp);
+    Value zero =
+        rewriter.create<arith::ConstantIndexOp>(op.getSrcMemref().getLoc(), 0);
+    Value newSubspan = rewriter.create<IREE::HAL::InterfaceBindingSubspanOp>(
+        op.getSrcMemref().getLoc(), subspanOp.getType(), subspanOp.getSet(),
+        subspanOp.getBinding(), subspanOp.getDescriptorType(), zero,
+        subspanOp.getDynamicDims(), subspanOp.getAlignmentAttr());
+    rewriter.restoreInsertionPoint(ip);
+
+    MLIRContext *context = rewriter.getContext();
+    AffineExpr sym0, sym1;
+    bindSymbols(context, sym0, sym1);
+    auto addMap = AffineMap::get(0, 2, {sym0 + sym1}, context);
+    auto divMap = AffineMap::get(0, 2, {sym0.floorDiv(sym1)}, context);
+
+    Value byteValue = rewriter.create<arith::ConstantIndexOp>(
+        op.getSrcMemref().getLoc(), numBytes.value());
+    // We assume that upper layers guarantee the byte offset is perfectly
+    // divisible by the element byte count so the content is well aligned.
+    Value offset = rewriter.create<AffineApplyOp>(
+        op.getLoc(), divMap, ValueRange{subspanOp.getByteOffset(), byteValue});
+
+    // Get the new index by adding the old index with the offset.
+    Value newIndex = rewriter.create<AffineApplyOp>(
+        op.getLoc(), addMap, ValueRange{op.getIndices().front(), offset});
+
+    rewriter.replaceOpWithNewOp<gpu::SubgroupMmaLoadMatrixOp>(
+        op, op.getType(), newSubspan, newIndex, op.getLeadDimension());
+
+    return success();
+  }
+};
+
+/// Folds the byte offset on subspan ops into the consumer mma store ops.
+struct FoldSubspanOffsetIntoMMAStore final
```

Comment:
Can this be improved a bit to avoid duplicated code? Looks like the whole function is the same except for getSrcMemref vs getDstMemref.
Would be nice to also share code with `FoldSubspanOffsetIntoLoadStore` if possible.

Reply:
I tried this with some logic like
if (std::is_same<OpType, gpu::SubgroupMmaStoreMatrixOp>::value) {
// mma store specific getDstMemref call
}
However, it wont build becuase the memref store and memref load dont have the same function, is that expected?

Reply:
> I tried this with some logic like
> if (std::is_same<OpType, gpu::SubgroupMmaStoreMatrixOp>::value) {
> // mma store specific getDstMemref call
> }
> However, it wont build becuase the memref store and memref load dont have the same function, is that expected?

In c++17 you can use `if constexpr` that should allow the code to not be instantiated if the if is false. Since we are now using c++17 this is the way to go. (Otherwise there are many other ways to refactor the code to avoid duplication by splitting things in functions :))